HungarianAlgorithm/Hungarian/Hungarian.cs

```diff
@@ -107,20 +106,16 @@ public Solution Solve()
                                     if (!eqGraph.ContainsEdge(well, s) && !eqGraph.ContainsEdge(s, well) &&
                                         cost == _graph.GetVertexLabel(well) + _graph.GetVertexLabel(s))
                                     {
-                                        var newEdge = new Edge<int>(s, well);
-                                        alternatingTree.AddEdge(newEdge);
-                                        eqGraph.AddEdge(newEdge);
-                                        nextT = edge.GetOtherVertex(s);
+                                        edgeToNextT = new Edge<int>(s, well);
+                                        eqGraph.AddEdge(edgeToNextT);
```

Comment:
Dobrze rozumiem, że tutaj wcale nie potrzebujemy powiększać `alternatingTree` za każdym razem jak rozszerzamy `eqGraph` bo tak właściwie interesuje nas później tylko znalezienie ścieżki powiększającej z `nextT` do `free_s` (jeśli wpadamy w ten wariant, że `nextT` nie jest incydentne z żadną krawędzią należącą do skojarzenia)? Tylko czy wtedy mamy pewność, że wszystkie krawędzie prowadzące do `free_s` są dodane do `alternatingTree` (w sensie czy zawsze istnieje chociaż jedna taka ścieżka w `alternatingTree`)? Czyli na przykład nie zachodzi przypadek, gdy zacieśniamy jednocześnie w danej iteracji dwie krawędzie i ścieżka z `nextT` do `free_s` przechodzi przez obie te krawędzie? Czy wtedy w `alternatingTree` istnieje może druga ścieżka?

Sory, że taki blok pytań, ale trochę głośno myślę w komentarzu :)

Reply:
Pytania bardzo trafne, bo nie przemyślałem tego tak dobrze :) i stąd się wziął błąd w 3_6_2

Odpowiadając trochę niepokolei:

> Dobrze rozumiem, że tutaj wcale nie potrzebujemy powiększać `alternatingTree` za każdym razem jak rozszerzamy `eqGraph` bo tak właściwie interesuje nas później tylko znalezienie ścieżki powiększającej z `nextT` do `free_s`?

Dokładnie tak

> Czyli na przykład nie zachodzi przypadek, gdy zacieśniamy jednocześnie w danej iteracji dwie krawędzie i ścieżka z `nextT` do `free_s` przechodzi przez obie te krawędzie? Czy wtedy w `alternatingTree` istnieje może druga ścieżka?

Zawsze będzie istniała ścieżka `free_s`-`nextT`.
Przykładowo: przypuśćmy, że w danej iteracji dodaliśmy do `eqGraph` 2 krawędzie: `free_s`->`t`, `s`->`nextT`.
Teraz zastanówmy się, w jaki sposób wierzchołek `s` dostał się do zbioru `S` (== do `alternatingTree`)? Został on dodany do drzewa równocześnie z jakimś innym wierzchołkiem `t_1` w którejś z poprzednich iteracji (wierzchołki do `S` dodawane są tylko w jednym miejscu - patrz: linia 138. niżej), ponieważ krawędź `t_1`->`s` należała do `eqGraph`. No a jeśli `t_1` należy do `T`, tzn. że istnieje jakaś `free_s`-`t_1`-ścieżka. Mamy więc ścieżkę `free_s`-`t_1`->`s`->`nextT`.
_Ergo_: każdy dodawany w skomentowanym przez Ciebie fragmencie wierzchołek `nextT` będzie osiągalny z `free_s` tylko na podstawie dotychczasowego drzewa, niezależnie od innych dodawanych krawędzi.

> Tylko czy wtedy mamy pewność, że wszystkie krawędzie prowadzące do `free_s` są dodane do `alternatingTree`?

I to jest właśnie mój błąd - nie mamy takiej pewności (choć wydaje mi się, że może o coś innego chodzić, niż masz tutaj na myśli) 🙂 Po kolei:
W zasadzie to nie tylko tak, że my _nie potrzebujemy_ powiększać `alternatingTree`, ale nawet _nie możemy_ tego zrobić więcej niż raz (dlatego to wyleciało poza blok - linia 116). Zauważmy, że w dalszej części sprawdzamy, czy `nextT` jest już skojarzone i na tej podstawie robimy jakieś akcje lub nie. Ale robimy je tylko dla tego jednego wierzchołka. Gdybyśmy dodali do zbioru `T` kilka wierzchołków na raz, to widniałyby one wszystkie jako sprawdzone i nigdy nie byłyby przetestowane.

Błąd w kodzie ma podobny powód, ale ukrył się w funkcji `AreAllNeighborsInSet`. Więcej napiszę w osobnym komentarzu